### PR TITLE
introduce integration ocm-additional-router

### PR DESCRIPTION
**Pull request by BumbleFeng**
https://issues.redhat.com/browse/APPSRE-3062

depends on https://gitlab.cee.redhat.com/service/app-interface/-/merge_requests/20784

Signed-off-by: Feng Huang <fehuang@redhat.com>

### CODE REVIEW

**Review by jfchevrette — APPROVED** [review]
This looks very clean. LGTM!

**Comment by maorfr** [comment]
following https://gitlab.cee.redhat.com/service/app-interface/-/merge_requests/20784#note_2556356, i'd consider renaming this integration to `ocm-additional-routers`?

File: reconcile/utils/ocm.py [hunk 1]
```diff
@@ -623,6 +623,63 @@ def delete_upgrade_policy(self, cluster, spec):
...
+            f'{CS_API_BASE}/v1/clusters/{cluster_id}/ingresses'
+        items = self._get_json(api).get('items')
+        if not items:
+            return results
+
+        for item in items:
+            # filter out default router
+            if item['default']:
+                continue
+            desired_keys = ['id', 'listening', 'dns_name', 'route_selectors']
```
Comment (Piojo) [inline reconcile/utils/ocm.py:647]: Move this out of this block, or you'll be generating this list over and over on each iteration. Ideally, make it a constant.

File: reconcile/ocm_additional_router.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    """ Sort diffs so we delete first and create later """
+    if diff['action'] == 'delete':
+        return 1
+    else:
+        return 2
+
+
+def act(dry_run, diffs, ocm_map):
+    diffs.sort(key=sort_diffs)
+    for diff in diffs:
```
Comment (Piojo) [inline reconcile/ocm_additional_router.py:78]: Please, don't use loop variables that differ from the iterated object by just one character. In this context, `for d in diffs` is perfectly understandable.

File: reconcile/ocm_additional_router.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+def sort_diffs(diff):
+    """ Sort diffs so we delete first and create later """
+    if diff['action'] == 'delete':
+        return 1
+    else:
+        return 2
+
+
+def act(dry_run, diffs, ocm_map):
```
Comment (Piojo) [inline reconcile/ocm_additional_router.py:76]: `act` is not a terribly useful name here. How about `update_routers`?

File: reconcile/ocm_additional_router.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    desired_state = []
+    for cluster in clusters:
+        cluster_name = cluster['name']
+        router = cluster['additionalRouter']
+        listening = 'internal' if router['private'] else 'external'
+        item = {
+            'listening': listening,
+            'cluster': cluster_name
+        }
+        selectors = cluster['additionalRouter']['route_selectors']
```
Comment (Piojo) [inline reconcile/ocm_additional_router.py:40]: This is `router['route_selectors']`, isn't it? Either use `router` all the way or drop this variable and use `cluster['additionalRouter']`. In any case, do the same things the same way always.

**Review by Piojo — CHANGES_REQUESTED** [review]
This is pretty big and meaningful, so I can haz tests? I don't need a very comprehensive suite, but I do need some degree of automated testing here. Feel free to reach out if you want help.

File: reconcile/ocm_additional_router.py (hunk 2 above)
Comment (maorfr) [reply reconcile/ocm_additional_router.py:78]: fwiw, this is very common across the code base, and specifically for ocm integrations

File: reconcile/ocm_additional_router.py (hunk 3 above)
Comment (maorfr) [reply reconcile/ocm_additional_router.py:76]: fwiw, `act` is a generic name for the different actions that we may perform, which differ between integrations

Comment (Piojo) [reply reconcile/ocm_additional_router.py:76]: ... but we already have `run` as our entry point for the "API", at some point we have to become specific, don't we?  Anyways, this isn't a big concern.

File: reconcile/test/test_ocm_additional_routers.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+        diffs = fixture['diffs']
+        integ.act(False, diffs, ocm_map)
+
+        ocm_act = fixture['ocm_act']
+
+        router_create = ocm_act['create']
+        expected = []
+        for c in router_create.keys():
+            expected.append(call(c, router_create[c]))
+        calls = ocm.create_additional_router.call_args_list
```
Comment (Piojo) [inline reconcile/test/test_ocm_additional_routers.py:110]: Why not:

```python
for c, route in router_create.items():
    ocm.create_additional_router.assert_called_with(c, route)
```
?

File: reconcile/test/test_ocm_additional_routers.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        self.assertEqual(calls, expected)
+
+        router_delete = ocm_act['delete']
+        expected = []
+        for c in router_delete.keys():
+            expected.append(call(c, router_delete[c]))
+        calls = ocm.delete_additional_router.call_args_list
```
Comment (Piojo) [inline reconcile/test/test_ocm_additional_routers.py:117]: As above.

**Review by Piojo — APPROVED** [review]
Awesome stuff. Thanks!

File: reconcile/test/test_ocm_additional_routers.py (hunk 5 above)
Comment (BumbleFeng) [reply reconcile/test/test_ocm_additional_routers.py:120]: assert_called_with can only get info of the last call.

Comment (Piojo) [reply reconcile/test/test_ocm_additional_routers.py:120]: `assert_any_call` is the right call here.

**Pull request merged**